src/main/java/zd/zero/waifu/motivator/plugin/assets/AssetManager.kt

```diff
@@ -0,0 +1,115 @@
+package zd.zero.waifu.motivator.plugin.assets
+
+import com.intellij.openapi.application.ex.ApplicationInfoEx
+import com.intellij.openapi.diagnostic.Logger
+import org.apache.commons.io.IOUtils
+import org.apache.http.client.config.RequestConfig
+import org.apache.http.client.methods.HttpGet
+import org.apache.http.impl.client.HttpClients
+import zd.zero.waifu.motivator.plugin.assets.LocalAssetService.hasAssetChanged
+import zd.zero.waifu.motivator.plugin.assets.LocalStorageService.createDirectories
+import zd.zero.waifu.motivator.plugin.assets.LocalStorageService.getLocalAssetDirectory
+import zd.zero.waifu.motivator.plugin.tools.toOptional
+import java.nio.file.Files
+import java.nio.file.Path
+import java.nio.file.Paths
+import java.nio.file.StandardOpenOption
+import java.util.*
+import java.util.concurrent.TimeUnit
+
+enum class AssetCategory(val category: String) {
+    VISUAL("visuals")
+}
+
+object AssetManager {
+    private const val ASSETS_SOURCE = "https://waifu.assets.unthrottled.io"
+
+    private val httpClient = HttpClients.custom()
+        .setUserAgent(ApplicationInfoEx.getInstance().fullApplicationName)
+        .build()
+
+    private val log = Logger.getInstance(this::class.java)
+
+    /**
+     * Will return a resolvable URL that can be used to reference an asset.
+     * If the asset was able to be downloaded on the local machine it will return a
+     * file:// url to the local asset. If it was not able to get the asset then it
+     * will return an https:// url to the remote asset.
+     */
+    fun resolveAssetUrl(assetCategory: AssetCategory, assetPath: String): String {
+        val remoteAssetUrl = constructRemoteAssetUrl(
+            assetCategory, assetPath
+        )
+        return constructLocalAssetPath(assetCategory, assetPath)
+            .flatMap {
+                resolveTheAssetUrl(it, remoteAssetUrl)
+            }.orElse(remoteAssetUrl)
+    }
+
+    private fun constructRemoteAssetUrl(
+        assetCategory: AssetCategory,
+        assetPath: String
+    ): String = "$ASSETS_SOURCE/${assetCategory.category}/$assetPath"
+
+    private fun resolveTheAssetUrl(localAssetPath: Path, remoteAssetUrl: String): Optional<String> =
+        if (hasAssetChanged(localAssetPath, remoteAssetUrl)) {
```

Comment:
Assets are check at most once a day. 

What happens is that I have a remote checksum and that is compared to the local checksum of the file on the machine. If they are different, then it is downloaded.

It costs me monies to send bytes to people, so I want to send as little as needed (32bytes is cheaper than 320,000,000bytes). Global CDNs cost that monies, but it's worth the user experience :)

Reply:
Oh my! maybe I'll set up a donation here? and I'll donate to it from myself hahaha, so we could track the money where it goes, let's collaboratively spend money so we can see our waifu manifest to our IDEs! 😂

Reply:
I wouldn't worry too much about it at the moment. My doki stuff is less than ~$5 a month, I spend more on other projects I support.
![image](https://user-images.githubusercontent.com/15972415/88730269-50d92c00-d0fb-11ea-8287-f435a81b55d6.png)